### ISSUE TITLE
Empty `type Query {\n}` in the sdl response

### ISSUE THREAD

**Issue by yuhanz** [apollographql/federation-jvm#23]
I encountered this issue when Apollo-Gateway has trouble recognizing sdl from a graphql service served from federation-jvm. The schema is copied from Apollo-Gateway sample for the inventory service, with a little modification by removing the `extend` keyword in the front and adding `@extends` directive:

```
  type Product @key(fields: "upc") @extends {
    upc: String! @external
    weight: Int @external
    price: Int @external
    shippingEstimate: Int @requires(fields: "price weight")
  }
```
This schema works well in the NodeJs version of graphql federation. However, with the same schema copied to the side of federation-jvm, Apollo-Gateway stopped to recognize the schema. The following message showed in the Apollo-Gateway log:
```
Encountered error when loading productsExt at http://localhost:9000/graphql: Syntax Error: Expected Name, found }
```

 Looking at the response body of sdl query from federation-jvm, the @extends was pulled in front of @keyword and there is an empty `type Query {\n}`.

```
{
  "data": {
    "_service": {
      "sdl": "type Product @extends @key(fields : \"upc\") {\n  price: Int @external\n  shippingEstimate: Int @requires(fields : \"price weight\")\n  upc: String! @external\n  weight: Int @external\n}\n\ntype Query {\n}\n"
    }
  }
}
```
The order of the `@extends` directive seemed to have no negative effect, as verified in by switching the order of `@extends` and `@keyword` in the NodeJS graphql federation. The empty `type Query {}` is probably the cause.

**Comment by zionts** [comment]
I believe this should be fixed by https://github.com/apollographql/federation-jvm/commit/2d7cc26f071a783a102fa5f40eb946367fc9f14f — can you check you are still having this problem on the latest version?

**Comment by sachindshinde** [comment]
Thanks for letting us know, @yuhanz ! We'll look into this issue more, and get back to you by March 9th.

**Comment by sachindshinde** [comment]
@yuhanz 
So we've determined the cause of the issue.

It turns out there's a bug in `graphql-java`'s `SchemaPrinter` in which it will use empty braces for object type definitions with no members (it also happens for interfaces, enums, and input objects). This violates the GraphQL grammar [as per spec](https://spec.graphql.org/June2018/#FieldsDefinition), and it should instead be using no braces at all. That is,
```graphql
type Query {
}
```
should really be
```graphql
type Query
```

This is why Apollo Gateway is giving you the error `Syntax Error: Expected Name, found }`; it uses `graphql.js` to parse, which fails because of the invalid grammar. The good news is that someone's already pushed a fix to `graphql-java` at graphql-java/graphql-java#1798 .

**Comment by yuhanz** [comment]
Great to know. Thanks for keep tracking. Looking forward to see a new build of graphql-java.

**Comment by sachindshinde** [comment]
@yuhanz 
In the meanwhile, we've copied their printer and patched it, so if you use `federation-jvm` v0.3.4 or later you shouldn't have any issues (note that using v0.3.4 requires upgrading to `graphql-java` v13).

I'll close out this issue, but let me know if you still experience similar issues and I can reopen.
